src/NServiceBus.Core/Routing/DistributionContext.cs

```diff
@@ -0,0 +1,49 @@
+namespace NServiceBus.Routing
+{
+    using System.Collections.Generic;
+    using Extensibility;
+    using Pipeline;
+
+    /// <summary>
+    /// The distribution context for the DistributionStrategy implementers.
+    /// </summary>
+    public class DistributionContext
+    {
+        /// <summary>
+        /// Creates a new distribution context.
+        /// </summary>
+        public DistributionContext(string[] receiverAddresses, OutgoingLogicalMessage message, string messageId, Dictionary<string, string> headers, ContextBag context)
+        {
+            ReceiverAddresses = receiverAddresses;
+            Message = message;
+            MessageId = messageId;
+            Headers = headers;
+            Context = context;
+        }
+
+        /// <summary>
+        /// The receiver addresses that can be taken into account for distribution.
+        /// </summary>
+        public string[] ReceiverAddresses { get; }
+
+        /// <summary>
+        /// The id of the outgoing message.
+        /// </summary>
+        public string MessageId { get; private set; }
+
+        /// <summary>
+        /// The headers of the outgoing message.
+        /// </summary>
+        public Dictionary<string, string> Headers { get; }
```

Comment:
Decided it is not worth going with readonly dictionary since for example error context doesn't have it either. Since it is a low level seam which requires high performance I don't see value in the interface dispatch there